File: services/calc/debrief_calc/tools/track_stats.py
"""
Track statistics tool.

Calculates statistics for a single track including:
- Number of positions
- Duration
- Total distance
- Average speed
"""

from __future__ import annotations

import math
import uuid
from typing import Any

from debrief_calc.models import (
    ContextType,
    GeoJSONFeatureDict,
    SelectionContext,
    ToolCategoryEnum,
    ToolParameter,
)
from debrief_calc.registry import tool
from debrief_schemas import OutputKindEnum


def _haversine_distance(lon1: float, lat1: float, lon2: float, lat2: float) -> float:
    """
    Calculate the great circle distance between two points in nautical miles.
    """
    # Convert to radians
    lon1, lat1, lon2, lat2 = map(math.radians, [lon1, lat1, lon2, lat2])

    # Haversine formula
    dlon = lon2 - lon1
    dlat = lat2 - lat1
    a = math.sin(dlat / 2) ** 2 + math.cos(lat1) * math.cos(lat2) * math.sin(dlon / 2) ** 2
    c = 2 * math.asin(math.sqrt(a))

    # Earth radius in nautical miles
    r = 3440.065

    return c * r


# Conversion factors from nautical miles
_NM_TO_KM = 1.852
_NM_TO_MI = 1.15078


def _convert_distance(distance_nm: float, unit: str) -> float:
    """Convert a distance from nautical miles to the requested unit."""
    if unit == "km":
        return distance_nm * _NM_TO_KM
    if unit == "mi":
        return distance_nm * _NM_TO_MI
    return distance_nm  # "nm" or default


def _convert_speed(speed_kts: float, unit: str) -> float:
    """Convert a speed from knots (nm/h) to the requested unit's per-hour equivalent."""
    if unit == "km":
        return speed_kts * _NM_TO_KM
    if unit == "mi":
        return speed_kts * _NM_TO_MI
    return speed_kts  # knots


def _speed_label(unit: str) -> str:
    """Return the speed unit label for the given distance unit."""
    if unit == "km":
        return "km/h"
    if unit == "mi":
        return "mph"
    return "kts"


def _distance_key(unit: str) -> str:
    """Return the statistics key for distance in the given unit."""
    return f"distance_{unit}"


def _speed_key(unit: str) -> str:
    """Return the statistics key for speed in the given unit."""
    labels = {"nm": "kts", "km": "kmh", "mi": "mph"}
    return f"average_speed_{labels.get(unit, 'kts')}"


def _calculate_track_stats(
    coordinates: list[list[float]], distance_unit: str = "nm"
) -> dict[str, Any]:
    """Calculate statistics from track coordinates."""
    if not coordinates:
        return {
            "point_count": 0,
            "duration_hours": 0,
            _distance_key(distance_unit): 0,
            _speed_key(distance_unit): 0,
        }

    point_count = len(coordinates)

    # Calculate total distance in nautical miles
    total_distance_nm = 0.0
    for i in range(1, len(coordinates)):
        prev = coordinates[i - 1]
        curr = coordinates[i]
        # Coordinates are [lon, lat, elevation?, time?]
        total_distance_nm += _haversine_distance(prev[0], prev[1], curr[0], curr[1])

    # Calculate duration if timestamps available (4th element)
    duration_hours = 0.0
    if len(coordinates[0]) >= 4 and len(coordinates[-1]) >= 4:
        start_time = coordinates[0][3]  # timestamp in ms
        end_time = coordinates[-1][3]
        duration_hours = (end_time - start_time) / (1000 * 60 * 60)

    # Calculate average speed in nm/h (knots)
    average_speed_kts = 0.0
    if duration_hours > 0:
        average_speed_kts = total_distance_nm / duration_hours

    # Convert to requested unit
    distance = _convert_distance(total_distance_nm, distance_unit)
    speed = _convert_speed(average_speed_kts, distance_unit)

    return {
        "point_count": point_count,
        "duration_hours": round(duration_hours, 2),
        _distance_key(distance_unit): round(distance, 2),
        _speed_key(distance_unit): round(speed, 2),
    }


@tool(
    name="track-stats",
    description="Calculate statistics for a single track including point count, duration, distance, and average speed",
    input_kinds=["TRACK"],
    output_kind=OutputKindEnum.trackSOLIDUSstatistics,
    context_type=ContextType.SINGLE,
    category=ToolCategoryEnum.calc,
    parameters=[
        ToolParameter(
            name="distance_unit",
            type="enum",
            description="Unit for distance measurements",
            choices=["nm", "km", "mi"],
            default="nm",
        )
    ],
)
def track_stats(context: SelectionContext, params: dict[str, Any]) -> list[GeoJSONFeatureDict]:
    """
    Calculate statistics for a single track.

    Args:
        context: SelectionContext with exactly one track feature (TRACK kind)
        params: Optional parameters (distance_unit)

    Returns:
        List containing one Feature with track statistics
    """
    distance_unit = params.get("distance_unit", "nm")
    if distance_unit not in ("nm", "km", "mi"):
        raise ValueError(f"distance_unit must be one of: nm, km, mi (got '{distance_unit}')")

    feature: GeoJSONFeatureDict = context.features[0]
    geometry: dict[str, Any] = feature.get("geometry", {})
    coordinates: list[list[float]] = geometry.get("coordinates", [])

    stats = _calculate_track_stats(coordinates, distance_unit)

    # Get the track's bounding box for the result geometry
    if coordinates:
        lons = [c[0] for c in coordinates]
        lats = [c[1] for c in coordinates]
        centroid = [sum(lons) / len(lons), sum(lats) / len(lats)]
    else:
        centroid = [0, 0]

    result_feature: GeoJSONFeatureDict = {
        "type": "Feature",
        "id": f"stats-{uuid.uuid4().hex[:8]}",
        "properties": {
            "source_track": feature.get("id", "unknown"),
            "source_name": feature.get("properties", {}).get("name", "unknown"),
            "statistics": stats,
        },
        "geometry": {"type": "Point", "coordinates": centroid},
    }

    return [result_feature]

File: apps/web-shell/src/tools/track/analysis/trackStats.ts
/**
 * Track Statistics tool — TypeScript implementation.
 * Calculates statistics for a single track: point count, duration,
 * total distance, and average speed.
 */

import type { MCPToolDefinition } from '../../../services/toolService';
import type { TrackFeature } from '@debrief/schemas';
import type { GeoJSONFeature } from '@debrief/utils';

const EARTH_RADIUS_NM = 3440.065;

function haversineDistanceNm(
  lon1: number, lat1: number, lon2: number, lat2: number,
): number {
  const dLat = ((lat2 - lat1) * Math.PI) / 180;
  const dLon = ((lon2 - lon1) * Math.PI) / 180;
  const a =
    Math.sin(dLat / 2) ** 2 +
    Math.cos((lat1 * Math.PI) / 180) *
    Math.cos((lat2 * Math.PI) / 180) *
    Math.sin(dLon / 2) ** 2;
  return 2 * EARTH_RADIUS_NM * Math.atan2(Math.sqrt(a), Math.sqrt(1 - a));
}

function generateUUID(): string {
  if (typeof crypto !== 'undefined' && crypto.randomUUID) return crypto.randomUUID();
  return 'xxxxxxxx-xxxx-4xxx-yxxx-xxxxxxxxxxxx'.replace(/[xy]/g, c => {
    const r = (Math.random() * 16) | 0;
    return (c === 'x' ? r : (r & 0x3) | 0x8).toString(16);
  });
}

export const toolDefinition: MCPToolDefinition = {
  name: 'track-stats',
  description:
    'Calculate statistics for a track: point count, duration, total distance, and average speed.',
  inputSchema: {
    type: 'object',
    properties: {
      features: { type: 'array', items: { type: 'object' } },
      params: {
        type: 'object',
        properties: {
          distance_unit: {
            type: 'string',
            enum: ['nm', 'km', 'mi'],
            default: 'nm',
            description: 'Unit for distance measurements',
          },
        },
      },
    },
  },
  annotations: {
    'debrief:selectionRequirements': [{ kind: 'TRACK', min: 1, max: 1 }],
    'debrief:category': 'track/analysis',
    'debrief:version': '1.0.0',
    'debrief:outputKind': 'track/statistics',
  },
};

type DistanceUnit = 'nm' | 'km' | 'mi';

// Conversion factors from nautical miles
const NM_TO_KM = 1.852;
const NM_TO_MI = 1.15078;

function convertDistance(distanceNm: number, unit: DistanceUnit): number {
  if (unit === 'km') return distanceNm * NM_TO_KM;
  if (unit === 'mi') return distanceNm * NM_TO_MI;
  return distanceNm;
}

function convertSpeed(speedKts: number, unit: DistanceUnit): number {
  if (unit === 'km') return speedKts * NM_TO_KM;
  if (unit === 'mi') return speedKts * NM_TO_MI;
  return speedKts;
}

function distanceKey(unit: DistanceUnit): string {
  return `distance_${unit}`;
}

function speedKey(unit: DistanceUnit): string {
  const labels: Record<DistanceUnit, string> = { nm: 'kts', km: 'kmh', mi: 'mph' };
  return `average_speed_${labels[unit]}`;
}

export function execute(
  features: TrackFeature[],
  params: Record<string, unknown>,
): GeoJSONFeature[] {
  const distanceUnit = ((params?.distance_unit as string) || 'nm') as DistanceUnit;
  if (!['nm', 'km', 'mi'].includes(distanceUnit)) {
    throw new Error(`distance_unit must be one of: nm, km, mi (got '${distanceUnit}')`);
  }

  const track = features.find(
    f => f.properties.kind === 'TRACK' && f.geometry.type === 'LineString',
  );
  if (!track) throw new Error('No track feature found');

  const coords = track.geometry.coordinates as number[][];
  if (coords.length === 0) throw new Error('Track has no coordinates');

  const pointCount = coords.length;

  // Total distance in nautical miles
  let totalDistanceNm = 0;
  for (let i = 1; i < coords.length; i++) {
    totalDistanceNm += haversineDistanceNm(
      coords[i - 1][0], coords[i - 1][1],
      coords[i][0], coords[i][1],
    );
  }

  // Duration from properties.times (epoch ms array, legacy wire format)
  const props = track.properties as unknown as Record<string, unknown>;
  let durationHours = 0;
  const times = props.times as number[] | undefined;
  if (times && times.length >= 2) {
    durationHours = (times[times.length - 1] - times[0]) / (1000 * 60 * 60);
  }

  const avgSpeedKts = durationHours > 0 ? totalDistanceNm / durationHours : 0;

  // Convert to requested unit
  const distance = convertDistance(totalDistanceNm, distanceUnit);
  const speed = convertSpeed(avgSpeedKts, distanceUnit);

  // Centroid for result feature placement
  const centroidLon = coords.reduce((s, c) => s + c[0], 0) / coords.length;
  const centroidLat = coords.reduce((s, c) => s + c[1], 0) / coords.length;

  const trackName = ((props.name ?? props.platform_name ?? track.id ?? 'Unknown') as string);

  return [{
    type: 'Feature',
    id: `stats-${generateUUID()}`,
    geometry: { type: 'Point', coordinates: [centroidLon, centroidLat] },
    properties: {
      kind: toolDefinition.annotations['debrief:outputKind'],
      name: `Statistics: ${trackName}`,
      source_track: String(track.id ?? ''),
      source_name: trackName,
      statistics: {
        point_count: pointCount,
        duration_hours: Math.round(durationHours * 100) / 100,
        [distanceKey(distanceUnit)]: Math.round(distance * 100) / 100,
        [speedKey(distanceUnit)]: Math.round(speed * 100) / 100,
      },
    },
  }];
}
